thewalrus/quantum.py

```diff
@@ -1,1300 +1,1320 @@
+        alphat = np.array(list(alpha) + ([0] * l))
+        x = 2 * alphat.real
+        p = 2 * alphat.imag
+        mu = np.concatenate([x, p])
+
+        tensor = state_vector(
+            mu,
+            cov,
+            normalize=False,
+            cutoff=cutoff,
+            hbar=2,
+            check_purity=False,
+            choi_r=choi_r,
+        )
+
+    if sf_order:
+        sf_indexing = tuple(chain.from_iterable([[i, i + l] for i in range(l)]))
+        return tensor.transpose(sf_indexing)
+
+    return tensor
+
+
+def probabilities(mu, cov, cutoff, parallel=False, hbar=2.0, rtol=1e-05, atol=1e-08):
+    r"""Generate the Fock space probabilities of a Gaussian state up to a Fock space cutoff.
+
+    .. note::
+
+        Parallization is already being done by OpenMP when calling ``density_matrix_element``.
+        To get a speed-up from using ``parallel=True`` it must be turned off by setting the
+        environment variable ``OMP_NUM_THREADS=1`` (forcing single threaded use). Remove the
+        environment variable or set it to ``OMP_NUM_THREADS=''`` to again use multithreading
+        with OpenMP.
+
+    Args:
+        mu (array): vector of means of length ``2*n_modes``
+        cov (array): covariance matrix of shape ``[2*n_modes, 2*n_modes]``
+        cutoff (int): cutoff in Fock space
+        parallel (bool): if ``True``, uses ``dask`` for parallelization instead of OpenMP
+        hbar (float): value of :math:`\hbar` in the commutation relation :math;`[\hat{x}, \hat{p}]=i\hbar`
+        rtol (float): the relative tolerance parameter used in `np.allclose`
+        atol (float): the absolute tolerance parameter used in `np.allclose`
+
+    Returns:
+        (array): Fock space probabilities up to cutoff. The shape of this tensor is ``[cutoff]*num_modes``.
+    """
+    if is_pure_cov(cov, hbar=hbar, rtol=rtol, atol=atol):  # Check if the covariance matrix cov is pure
+        return np.abs(state_vector(mu, cov, cutoff=cutoff, hbar=hbar, check_purity=False)) ** 2
+    num_modes = len(mu) // 2
+
+    if parallel:
+        # save old env variable (None if empty) and set single-thread use in OpenMP
+        OMP_NUM_THREADS = os.getenv("OMP_NUM_THREADS")
+        os.environ["OMP_NUM_THREADS"] = "1"
+
+        compute_list = []
+        # create a list of parallelizable computations
+        for i in product(range(cutoff), repeat=num_modes):
+            compute_list.append(dask.delayed(density_matrix_element)(mu, cov, i, i, hbar=hbar))
+
+        probs = np.maximum(
+            0.0, np.real_if_close(dask.compute(*compute_list, scheduler="processes"))
+        ).reshape([cutoff] * num_modes)
+        # maximum is needed because sometimes a probability is very close to zero from below
+
+        # restore env variable to value before (or remove if it wasn't set)
+        if OMP_NUM_THREADS:
+            os.environ["OMP_NUM_THREADS"] = OMP_NUM_THREADS
+        else:
+            del os.environ["OMP_NUM_THREADS"]
```

Comment:
Nice! I only have one (minor) concern --- this is modifying a global environment variable.

If the user is running multiple Python scripts at the same time, or maybe even any other program, then this code-block will cause side effects in the other running processes

Reply:
I agree that this could be an issue. I would say that this is OK if conveyed clearly to the user (e.g. by mentioning it in the docstring). Another option would be to simply avoid changing the environment variable in the function itself and only do it during testing. Then it would be up to the user to switch off OpenMP parallelisation before running with `parallel=True`, although this could still pose a similar issue since programs might freeze/crash if not taking this into account.

Reply:
The more I think about it, it definitely feels a bit strange for a Python library to be changing a users environment variables. If the calculation crashes midway, for instance, the environment variables will never be reset to the original values.

Reply:
yeah, I think it makes sense to do it in the tests, but otherwise you tell the uer how to do it in the docstring.

Reply:
I'm leaning towards this being something we communicate clearly in the documentation, but don't actually modify ourselves.

Unless, this is a common occurrence? Are there examples of other python libraries modifying `OMP_NUM_THREADS`?

Reply:
I agree that it's a bit strange, and perhaps a quite bad thing to do. I'll remove it from the function then and simply change the environment variable in the tests (would this still be OK?), and also updating the docstring.

I don't know of any other places modifying `OMP_NUM_THREADS`. :thinking: